experimental/dds/stack/src/sharedStack.ts

```diff
@@ -0,0 +1,162 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IEvent } from "@fluidframework/common-definitions";
+import { IsoBuffer } from "@fluidframework/common-utils";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { IChannel, IChannelAttributes, IChannelFactory, IChannelServices, IChannelStorageService, IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ITelemetryContext, ISummaryTreeWithStats, IGarbageCollectionData } from "@fluidframework/runtime-definitions";
+import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import { IFluidSerializer, SharedObject } from "@fluidframework/shared-object-base";
+
+/**
+ * Events useful for testing and debugging a {@link SharedStack}
+ */
+export interface DiagnosticEvents extends IEvent {
+    (event: "downloadedBlob", handler: () => void);
+    (event: "uploadedBlob", handler: () => void);
+    (event: "processedOp", handler: () => void);
+}
+
+/**
+ * A stack DDS which stores every element in its own attachment blob.
+ */
+export class SharedStack<T = unknown> extends SharedObject<DiagnosticEvents> {
+    private head?: IFluidHandle<ArrayBufferLike>;
+
+    /** Construct a new SharedStack */
+    public constructor(
+        id: string,
+        runtime: IFluidDataStoreRuntime,
+        attributes: IChannelAttributes,
+        telemetryContextPrefix: string
+    ) {
+        super(id, runtime, attributes, telemetryContextPrefix);
+    }
+
+    /** True iff there are no elements pushed to this stack */
+    public isEmpty(): boolean {
+        return this.head === undefined;
+    }
+
+    /** Download, remove and return the top element on this stack */
+    public async pop(): Promise<T> {
+        if (this.head === undefined) {
+            throw new Error("SharedStack is empty");
+        }
+
+        const head = await get<Node<T>>(this.head, this.serializer);
```

Comment:
If two clients call pop() simultaneously, I think this will result in them both receiving the same element.  Is that OK/expected for the API contract you want to provide?

Reply:
Yes, I think that's okay. Pop is a little weird since it is both a read and a write simultaneously. I think the important part is that even if multiple clients pop at the same time and read the same element, their writes (setting the head to popped.next) don't conflict since they are identical, and the resulting state will be identical regardless of which order they are applied in.

Reply:
I think this might surprise some customers, though I'd be ok with it being a design choice -- probably at least document it heavily (maybe consider a section in the readme "How is SharedStack alike/different from the stacks you're used to?").

Reply:
I imagine the alternative is for pop() to be an ack-based modification of the data (consensus-like again).  I think most customers of a stack probably want it to be speedy though, so this might be getting away from the value prop opportunity.

Reply:
Right, agreed. I should be more clear here that I don't really intend anyone to use this DDS for a real project; this package is here primarily to demonstrate some of the shortcomings of our current GC APIs (which will be in the next PR I make here), as well as give a minimal test bed for interacting with the attachment blob APIs. This is the reason it's in the experimental folder.

An alternative is to just keep this on a branch or a fork and not check it into the repo, but then it runs the risk of getting stale over time as framework APIs and contracts change. I hope I am not using the experimental folder incorrectly, but it seemed to me that this kind of thing is what it's for (so long as what is checked in meets at least some bar of quality, which I want it to, and you are helping with that - thanks :) )

Reply:
No worries -- I personally don't use the experimental folder and think we should get rid of it (makes layer-checking ineffective and no one looks at our folder structure externally so it's an ineffective hint).  The things I care most about are (1) does the package publish anywhere and (2) what's its scope.

Given that purpose you mention, a couple options (others might have different opinions, but I'd be fine with any of these):
- Keep this as `@fluid-experimental/stack` and make it `"private": true` -- I don't particularly mind if it's in the experimental directory or the main dds directory.
    - `@fluid-experimental` is a hint that it's an in-development feature that will eventually be released to the public, generally.  Do this if you think you'll eventually want customers to try it out.
- Switch it to `@fluid-internal/stack` if you never intend to graduate it to a public package.
- Move it into some appropriate existing `@fluid-internal` test package if you'd characterize it as a test tool only.
- Include it in whatever package you create for demonstrating the API shortcomings (as a non-exported member) -- I'm assuming some `@fluid-example` package.  Probably the best option if you don't need to import it into multiple consumers.